pkg/operator/operator_test.go

```diff
@@ -0,0 +1,165 @@
+	)
+	apiV1 := client.CertificatesV1().CertificateSigningRequests()
+	// Fire off a goroutine to mock asynchronous kube-apiserver.
+	go func() {
+		// Block until kube client has newly-created CSR.
+		if err := wait.Poll(50*time.Millisecond, timeout, func() (bool, error) {
+			var gerr error
+			// Ignore any errors coming from get API call as
+			csr, gerr = apiV1.Get(ctx, fqdn, metav1.GetOptions{})
+			if apierrors.IsNotFound(gerr) {
+				return false, nil
+			}
+			// Check if certificate has been approved by the signing function.
+			if len(csr.Status.Conditions) < 1 {
+				return false, nil
+			} else {
+				return csr.Status.Conditions[0].Type == v1.CertificateApproved, nil
+			}
+		}); err != nil {
+			t.Errorf("timeout waiting for CSR: %s", err)
+		}
+
+		// Mock the kube-apiserver issuing the certificate by writing toy data to status.
+		// This will unblock the keypair generation goroutine.
+		csr.Status.Certificate = certBytes
+		if _, err := apiV1.UpdateStatus(ctx, csr, metav1.UpdateOptions{}); err != nil {
+			t.Errorf("updating csr while issuing cert: %s", err)
+		}
+	}()
+}
+
+func TestEnsureCerts(t *testing.T) {
+	var (
+		timeout     = 3 * time.Second
+		ctx, cancel = context.WithTimeout(context.Background(), timeout)
+		certBytes   = []byte{1, 2, 3, 4}
+	)
+	t.Cleanup(cancel)
+
+	dir, err := ioutil.TempDir("", "temp")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.RemoveAll(dir) // clean up
+
+	client := fake.NewSimpleClientset()
+	mockKubeAPIServer(ctx, client, t)
+
+	for _, tc := range []struct {
+		desc       string
+		opts       Options
+		expectCert string
+		expectKey  string
+		expectErr  bool
+	}{
+		{
+			desc:       "input key/cert",
+			opts:       Options{Key: "a2V5", Cert: "Y2VydA=="},
+			expectCert: "cert",
+			expectKey:  "key",
+			expectErr:  false,
+		},
+		{
+			desc:      "bad cert",
+			opts:      Options{Cert: "not a cert", Key: "not a key"},
+			expectErr: true,
+		},
+		{
+			desc:       "self generate keys/cert",
+			opts:       Options{Cert: "", Key: ""},
+			expectCert: "1,2,3,4",
+			expectErr:  false,
+		},
+		{
+			desc:      "cert and no key",
+			opts:      Options{Cert: "cert", Key: ""},
+			expectErr: true,
+		},
+		{
+			desc:      "key and not cert",
+			opts:      Options{Cert: "", Key: "key"},
+			expectErr: true,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			op := Operator{
+				opts:       tc.opts,
+				kubeClient: client,
+			}
+			_, err := op.ensureCerts(ctx, dir)
+			if err != nil {
+				if !tc.expectErr {
+					t.Fatalf("want err: %v; got %v", tc.expectErr, err)
+				}
+				return
+			}
+			// Test outputed files.
+			outCert, err := ioutil.ReadFile(path.Join(dir, "tls.crt"))
+			if err != nil {
+				t.Fatalf("error reading from cert file: %v", err)
+			}
+			outKey, err := ioutil.ReadFile(path.Join(dir, "tls.key"))
+			if err != nil {
+				t.Fatalf("error reading from key file: %v", err)
+			}
+			if len(op.opts.Cert) == 0 && len(op.opts.Key) == 0 {
```

Comment:
So at this point there is different validation logic for some test cases as well as a mock server only used by one of them. I think at this point the table-driven test loses it's benefit and I'd just split them into two tests, one for server-signed ones and one for explicit ones.